dttools/src/jx_repl.c

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright (C) 2021- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+/*
+This is a program to explore the jx library.
+It reads in a JX expression, evaluates it, prints out the result, then saves it to the context.
+Results of previous expressions can be referenced via the `out_%d` symbol, and their query via `in_%d`
+The program exits once EOF is reached or after the user enters the `quit` or `exit` command
+*/
+
+#include "jx.h"
+#include "jx_sub.h"
+#include "jx_eval.h"
+#include "jx_parse.h"
+#include "jx_print.h"
+#include "jx_pretty_print.h"
+
+#include <stdio.h>
+#include <string.h>
+
+#ifdef HAS_LIBREADLINE
+#include "readline/readline.h"
+#include "readline/history.h"
+#endif
+
+#define CATALOG_URL "http://catalog.cse.nd.edu:9097/query.json" 
+#define MAX_LINE 200
+
+
+const char * MSG_WELCOME =
+    "Welcome to the JX Language Explorer.\n"
+    "\n"
+    "Type 'help' for help\n"
+    "\n";
+
+const char * MSG_HELP =
+    "\n"
+    "  help          display this message\n"
+    "  functions     display a list of functions supported by the JX language\n"
+    "  values        display a list of values supported by the JX language\n"
+    "  operators     display a list of operators supported by the JX language\n"
+    "  in_#          the #'th input query\n"
+    "  out_#         result of in_#\n"
+    "  catalog       alias to fetch catalog data\n"
+    "  quit|exit     exit program\n"
+    "\n";
+
+const char * MSG_FUNCTIONS =
+    "\n"
+    "  format( \"str: %%s int: %%d float: %%f\", \"hello\", 42, 3.14159 )\n"
+    "  join( array, delim )\n"
+    "  range( start, stop, step )\n"
+    "  ceil( value )\n"
+    "  floor( value )\n"
+    "  basename( path )\n"
+    "  dirname( path )\n"
+    "  escape( string )\n"
+    "  len( array )\n"
+    "  fetch( URL/path )\n"
+    "  select( boolean, array )\n"
+    "  project( expression, array )\n"
+    "  schema( object )\n"
+    "  like( string, object )\n"
+    "\n";
```

Comment:
This is a nice help text, but we should find a way to keep it synchronized with the list of available functions in jx_exal / jx_function.  Maybe we just need some comments in both places to remember?